<html>
<head>
    <title>QUAST 2.2 manual</title>
    <style type="text/css">
        body {
            margin-top: 30px;
            margin-left: 50px;
            max-width: 800px;
            font-family: Tahoma,sans-serif;
            font-size: 14px;
        }
        pre.code {
            background-color: #EEE;
            padding: 5px 0px;
            margin-top: 5px;
            margin-bottom: 15px;
        }
        table td {
            vertical-align: top;
        }
        h2, h3, h4 {
            margin-bottom: -10px;            
        }
        h2 {
            margin-top: 40px;
        }
        h3 {
            margin-top: 50px;
        }
        h4 {
            margin-top: 30px;
            font-size: 1.1em;
        }
        ul {
            margin-top: -12px;
        }
        ul li, ol li {
            margin-bottom: 5px;
        }
        .options {
            margin-left: 34px;
        }
        .option {      
            margin-top: 15px;
            margin-bottom: 5px;
        }
        .options .option:first-child {
            margin-top: 4px;
        }
        .metric-name, .metric-ref {
            font-family: Georgia,serif;            
        }
        .metric-name {
            font-weight: bold;
        }
        .metrics_description p {
            margin-bottom: 25px;
        }
        .hs { /* 10<span class="hs"></span>000 */
            margin-left: .2em; 
        }
        .rhs { /* 1<span class="rhs">&nbsp;</span>kb */            
            font-size: 50%;
            line-height: 1;
        }
    </style>
</head>
<body>
<h1>QUAST 2.2 manual</h1>

<p> QUAST stands for <u>QU</u>ality <u>AS</u>sesment <u>T</u>ool. 
    The tool evaluates genome assemblies by computing various metrics.
    <br>
    <br>
    You can find all project news and the latest version of the tool
    at <a href="http://sourceforge.net/projects/quast/">http://sourceforge.net/projects/quast</a>.
    <br>
    <br>
    QUAST utilizes <a href="http://mummer.sourceforge.net/">MUMmer</a>, <a href="http://exon.gatech.edu/genemarks.cgi">GeneMarkS</a>, <a href="http://exon.gatech.edu/genemarks.cgi">MetaGeneMark</a>, 
    <a href="http://cbcb.umd.edu/software/glimmerhmm/">GlimmerHMM</a>
    and <a href="http://gage.cbcb.umd.edu/index.html">GAGE</a>. These tools are built in, so you don't need to install
    them separately.
    <br>
    <br>
    Version 2.2 of QUAST was released under GPL v2 (see <a href="LICENSE">LICENSE</a> for details) on 3 July 2013.
</p>

<h2>Contents</h2>
<ol>
  <li><a href="#sec1">Installation</a></li> 
  <li><a href="#sec2">Running QUAST</a> 
    <ol>
      <li><a href="#sec2.1">For impatient people</a></li>
      <li><a href="#sec2.2">Input data</a></li>
      <li><a href="#sec2.3">GAGE mode</a></li>
      <li><a href="#sec2.4">Command line options</a></li>
      <li><a href="#sec2.5">Metagenomic assemblies</a></li>
    </ol>
  </li>
  <li><a href="#sec3">QUAST output</a> 
    <ol>
      <li><a href="#sec3.1">Metrics description</a>
        <ol>
          <li><a href="#sec3.1.1">Summary report</a></li>
          <li><a href="#sec3.1.2">Misassemblies report</a></li>
          <li><a href="#sec3.1.3">Unaligned report</a></li>
        </ol>
      </li>
      <li><a href="#sec3.2">Plots descriptions</a></li>      
    </ol>
  </li>
  <li><a href="#sec4">Adjusting QUAST reports and plots</a></li> 
  <li><a href="#sec5">Citation</a></li>
  <li><a href="#sec6">Feedback and bug reports</a></li> 
</ol>

<a name="sec1"></a>
<h2>1. Installation</h2>

<p>
    QUAST can be run on Linux or Mac OS.
</p>
<p>
    It requires:
</p>
<ul>
    <li>python 2 (2.5 or higher)</li>
    <li>perl 5.6.0 or higher</li>
    <li>g++</li>
    <li>make</li>
    <li>sh</li>
    <li>csh</li>
    <li>sed</li>
    <li>awk</li>
    <li>ar</li>    
</ul>
All those tools are usually preinstalled on Linux.<br>
Mac OS, however, initially misses <code>make</code>, <code>g++</code> and <code>ar</code>, so you will have to install 
<a href='https://developer.apple.com/xcode/'>Xcode</a> (or only
<a href='https://developer.apple.com/downloads/index.action?name=Command%20Line%20Tools'>Command Line Tools for Xcode</a>) to make them available.
<br>
<br>
It is also highly recommended to install the <a href="http://matplotlib.sourceforge.net/">Matplotlib</a> Python library for drawing plots.
<br>
It can be done with Python <a href="http://www.pip-installer.org/">pip-installer</a>:
<pre class="code">
    pip install matplotlib
</pre>
Or with the <a href="http://peak.telecommunity.com/DevCenter/EasyInstall">Easy Install</a> Python module:
<pre class="code">
    easy_install matplotlib
</pre>
Or on Ubuntu by typing:
<pre class="code">
    sudo apt-get install python-matplotlib
</pre>
<br>
To download the <a href="https://downloads.sourceforge.net/project/quast/quast-2.2.tar.gz">QUAST source code tarball</a> and extract it, type:
<pre class="code">
    wget https://downloads.sourceforge.net/project/quast/quast-2.2.tar.gz
    tar -xzf quast-2.2.tar.gz
    cd quast-2.2
</pre>

<a name="sec2"></a>
<h2 style='margin-bottom: -40px;'>2. Running QUAST</h2>
<a name="sec2.1"></a>
<h3>2.1 For impatient people</h3><br>
Running QUAST on test data from the installation tarball (reference genome, gene and operon annotations, and two assemblies of the first 10<span class="rhs">&nbsp;</span>kbp of <i>E.&nbsp;coli</i>):
<pre class="code">
    python quast.py test_data/contigs_10k_1.fasta test_data/contigs_10k_2.fasta \
        -R test_data/reference_10k.fasta.gz \
        -G test_data/genes_10k.txt -O test_data/operons_10k.txt
</pre>    
View the summary of the evaluation results with the <a href="http://www.greenwoodsoftware.com/less/">less</a> utility:
<pre class="code">
    less quast_results/latest/report.txt
</pre>    

<a name="sec2.2"></a>
<h3>2.2 Input data</h3>
<p>
    The <code>test_data</code> directory contains examples of assembly, reference, gene and operon files.<br>
    <br>
    <b>Sequences</b><br>
    The tool accepts assemblies and references in FASTA format. Files may be compressed with zip, gzip, or&nbsp;bzip2.<br>
    Multiple reference chromosomes can be provided as separate sequences in a single FASTA file.<br>
    <span style='line-height: 50%;'>&nbsp;</span><br>  
    Maximum assembly length is 4.29<span class="rhs">&nbsp;</span>Gbp.<br>
    Maximum length of a reference sequence is 536<span class="rhs">&nbsp;</span>Mbp. The number of sequences is not limited.<br>
    <span style='line-height: 50%;'>&nbsp;</span><br>  
    Those restrictions belongs to Nucmer, a tool that QUAST applies to align contigs to a reference genome. 
    The metrics that do not require alignment are computed in any case.<br>
    <br>
    <b>Genes and operons</b><br>
    One can also specify files with gene and operon positions in the reference. QUAST will count fully and partially aligned regions, 
    and output <a href='#genes'>total values</a> and <a href='#gene_plot'>comulative plots</a>.<br>
    <span style='line-height: 50%;'>&nbsp;</span><br>  
    The following file formats are suported:
    <ul>
        <li>GFF, versions <a href="http://www.sanger.ac.uk/resources/software/gff/spec.html">2</a> and <a href="http://www.sequenceontology.org/gff3.shtml">3</a>
        (note: &lt;feature&gt; field is either "gene" or "operon");
        <li>the <a href="http://www.ncbi.nlm.nih.gov/gene">format used by NCBI</a> for genes ("Summary (text)");
        <li>four tab-separated columns: sequence name, gene/operon id, start position, end postion.
    </ul>
    Note that the sequence name has to match a name in the reference file.<br>
    <span style='line-height: 50%;'>&nbsp;</span><br>  
    Coordinates are 1-based, i.e. the first nucleotide in the reference has position 1, not 0.
    If a <i>start postion</i> less than a corresponding <i>end postion</i>, such gene or operon is on a positive strand, otherwise it is on a negative strand.
</p>    

<a name="sec2.3"></a>
<h3>2.3 GAGE mode</h3>
<p>
    <a href="http://gage.cbcb.umd.edu/index.html">GAGE</a> is a well-known assessment tool. However, it has limitations:
    <ul>
        <li>Only one assembly per run. It complicates assembly comparison. 
        <li>Fixed threshold for a minimum contig length (200<span class="rhs">&nbsp;</span>bp).
    </ul>
    These issues are solved by QUAST in GAGE mode (run with a <code>--gage</code> option). QUAST filters contigs according to a specified threshold and runs GAGE on every assembly. 
    GAGE statistics (see the <a href="http://gage.cbcb.umd.edu/index.html">GAGE site</a> and the <a href="http://genome.cshlp.org/content/early/2012/01/12/gr.131383.111">GAGE paper</a> for descriptions) 
    are reported in addition to a standard QUAST report.<br>
    <span style='line-height: 50%;'>&nbsp;</span><br>  
    Note:<br><br>
    <ul>
        <li>GAGE requires a reference genome.</li>
        <li>Java and Java Compiler must be installed on your machine. Tested with OpenJDK 6.</li>
    </ul>
</p>

<a name="sec2.4"></a>
<h3>2.4 Command line options</h3>
<br>
QUAST runs from a command line as follows:
<pre class='code'>
    python quast.py [options] &lt;contig_file(s)&gt;
</pre>

Options:
<div class='options'>
<div class='option'>
<code><b>-o</b> &lt;output_dir&gt;</code>
</div>
Output directory. The default value is <code>quast_results/results_&lt;date_time&gt;</code>.<br>
Also, a symlink <code>quast_results/latest</code> is created.<br>
<br>
Note: QUAST reuses Nucmer alignments if run repeatedly on the same directory.
Thus, you can efficiently reuse already computed results when running QUAST with different parameters, or adding more assemblies to an existing comparison.

<div class='option'>
<code><b>-R</b> &lt;path&gt;</code>
</div>
Reference genome file. Optional. Many metrics can't be evaluated without a reference. If this is omitted, QUAST will only report the metrics that can be evaluated without a reference.

<div class='option'>
<code><b>-G</b> &lt;path&gt;</code> (or <code>--genes &lt;path&gt;</code>)
</div>
File with gene positions in reference. See details about the file format in <a href="#sec2.2">section 2.2</a>.<br>
<span style='line-height: 50%;'>&nbsp;</span><br>
If you do not have gene positions, you can make QUAST predict genes by the <code><a href='#gene_finding'>--gene-finding</code> option</a>.<br>

<div class='option'>
<code><b>-O</b> &lt;path&gt;</code> (or <code>--operons &lt;path&gt;</code>)
</div>
File with operon positions in reference. See details about the file format in <a href="#sec2.2">section 2.2</a>

<div class='option'>
<a name='min_contig'></a>
<code><b>&#8209;&#8209;min&#8209;contig</b> &lt;int&gt;</code>
</div>
Lower threshold for a contig length. Shorter contigs won't be taken into account
(except for some metrics, see <a href="#sec3">section 3</a>). The default value is 500.

<div class='option'>
<code><b>-t</b></code> (or <code>--threads</code>) <code>&lt;int&gt;</code>
</div>
Maximum number of threads. The default value is the number of CPUs. If QUAST fails to determine the number of CPUs, the number is set to 4.

<div class='option'>
<code><b>--labels</b></code> (or <code>-l</code>) <code>&lt;label,label...&gt;</code>
</div>
Human-readable assembly names. Those names will be used in reports, plots and logs. For example:<br>
<div style='margin-left: 30px; margin-top: 5px; margin-bottom: 10px;'>
<code>-l SPAdes,IDBA-UD</code>  
</div>
If your labels include spaces, use quotes:<br>
<div style='margin-left: 30px; margin-top: 5px; margin-bottom: 10px;'>
<code>-l SPAdes,"Assembly 2",Assembly3</code>
</div>
<div style='margin-left: 30px; margin-top: 5px; margin-bottom: 10px;'>
<code>-l "SPAdes 2.5, SPAdes 2.4, IDBA-UD"<br></code>
</div>

</div>

<br>
Advanced options:
<div class='options'>
<div class='option'>
    <code><b>--scaffolds</b></code>
</div>
The assemblies are scaffolds (rather than contigs). QUAST will add split versions
of assemblies to the comparison. Assemblies are split by continuous fragments of N's of length &ge; 10.

<div class='option'>
    <code><b>--ambiguity-usage</b> &lt;<b>none</b>|<b>one</b>|<b>all</b>&gt;</code>
</div>
Way of processing equally good alignments (probably repeats):<br>
<table style="margin-left: 30px; font-size: 1em; vertical-align: bottom;">
    <tr><td style='width: 40px;'><code>none</code></td><td>skip all such alignments;</td></tr>
    <tr><td><code>one</code></td><td>take only one (the first one);</td></tr>
    <tr><td><code>all</code></td><td>use all alignements. Can cause a significant increase of <span class='metric-ref'># mismatches</span>
    (repeats are almost always inexact due to accumulated SNPs, indels, etc.).</td></tr>
</table>
<div class='option'>
    <a name='gene_finding'></a><code><b>--gene-finding</b></code>
</div>
Enables gene finding. Affects perfomance, thus disabled by default.<br>
<span style='line-height: 50%;'>&nbsp;</span><br>  
By default, we asume that the genome is prokaryotic, and apply GenemarkS for gene finding.
If the genome is eukaryotic, add the <a href='#eukaryote'><code>--eukaryte</code></a> option to enable GlimmerHMM instead.
If it is a metagenome, add the <a href='#meta'><code>--meta</code></a> option.<br>
<span style='line-height: 50%;'>&nbsp;</span><br>  
If a <a href='#sec2.2'>gene file</a> is provided by <code>-G</code> as well, both 
<span class='metric_ref'><a href='#genes'># genes</a></span> in the file covered by the assembly, and
<span class='metric_ref'><a href='#predicted_genes'># predicted genes</a></span> are reported. Note that operons are not predicted,
but a file of known operon positions can be provided instead.<br>

<div class='option'>
    <code><b>--gene-thresholds</b> &lt;int,int,...&gt;</code>
</div>
Comma-separated list of thresholds for gene lengths to find with a finding tool. The default value is 0,300,1500,3000.
Note: this list is used only if <code>--gene-finding</code> option is specified.

<div class='option'>
    <a name="eukaryote"></a>
    <code><b>--eukaryote</b></code>
</div>
Genome is eukarytic. Affects gene finding and contig alignment:<br>
<ol class='my_ol' style='margin-top: 0;'>
<li>For prokaryotes (which is default), GenemarkS is used. For eukaryotes, GlimmerHMM is used.
<li>By default, QUAST assumes that a genome is circular and correctly processes its linear representation.
This options indicates that the genome is not circular.
</ol>

<div class='option'>
    <a name="meta"></a>
    <code><b>--meta</b></code>
</div>
Use MetaGeneMark for gene finding, if the <code>--gene-finding</code> option is specified.
If the <code>--eukaryote</code> option is also provided, MetaGeneMark still will be used.<br>
<span style='line-height: 50%;'>&nbsp;</span><br>  
Note: if you have multiple references, <a href='#sec2.5'>use metaquast.py</a> instead
(it is in the same directory as quast.py).

<div class='option'>
    <code><b>--use-all-alignments</b></code>
</div>
Compute <span class='metric-ref'>genome fraction, # genes, # operons</span> metrics in the manner used in QUAST v.1.*.
By default, QUAST v.2.* filters out ambiguous and redundant alignments, keeping only one alignment per contig 
(or one set of non-overlapping or slightly-overlapping alignments). This option makes QUAST count all alignments.

<div class='option'>
    <code><b>--contig-thresholds</b> &lt;int,int,...&gt;</code>
</div>
Comma-separated list of contig length thresholds. Used in <span class='metric-ref'># contigs &ge; x</span> and
<span class='metric-ref'>total length (&ge; x)</span> metrics (see <a href="#sec3">section 3</a>). The default value is 0,1000.

<div class='option'>
    <code><b>--est-ref-size</b> &lt;int&gt;</code>
</div>
Estimated reference size (in bases) for computing <span class='metric-ref'>NGx</span> statistics. This value will be used only if a reference genome file is
not specified (see <code>--R</code> option).

<div class='option'>
    <code><b>--strict-NA</b></code>
</div>
Break contigs at every misassembly event (including local ones) to compute <span class='metric-ref'>NAx</span> and
<span class='metric-ref'>NGAx</span> statistics. By default, QUAST breaks contigs <i>only at extensive</i> misassemblies (not local ones).

<div class='option'>
    <code><b>--gage</b></code>
</div>
Starts QUAST in "GAGE mode" (see <a href="#sec2.3">section 2.3</a>).
Note: in this case, you also have to set the <code>-R</code> option.

<div class='option'>
    <code><b>--test</b></code>
</div>
Runs the tool on a data from the <code>test_data</code> folder.

<div class='option'>
    <code><b>-h</b></code> (or <code>--help</code>)
</div>
Prints help.

</div>

<a name="sec2.5"></a>
<h3>2.5 Metagenomic assemblies</h3>
<p>
    The <code>metaquast.py</code> script accepts multiple references. One can provide several files, a merged FASTA file with multiple sequences, or a combination.
    The tool partitions all contigs in groups aligned to each reference.
    Then it runs quast.py several times:<br>
    <ul>
        <li>for all references in combination,
        <li>for each reference separately, by using corresponding contigs,
        <li>for the rest of the contigs that were not aligned anywhere.
    </ul>       

<!-- <table style="margin-left: 20px;">
    <tr>
        <td style="padding-right: 20px; padding-bottom: 5px;">- for all references in combination,</td>
        <td></td>
    </tr>
    <tr>
        <td style="padding-right: 20px; padding-bottom: 5px;">- for each reference separately, by using corresponding contigs,</td>
        <td></td>
    </tr>
    <tr>
        <td style="padding-right: 20px; padding-bottom: 5px;">- for the rest of the contigs that were not aligned anywhere.</td>
        <td></td>
    </tr>
</table> -->

<p>All outputs are in separate directories inside the directory provided by <code>-o</code> (or in quast_results/latest).</p>
Usage:
<pre class="code">
    python metaquast.py contigs_1 contigs_2 ... -R reference_1,reference_2,reference_3,...
</pre>

All options are the same as for quast.py, except for <code>-R</code>: it can accept multiple references.</p>

<a name="sec3"></a>
<h2>3. QUAST output</h2>
<p>If an output path was not specified manually, QUAST puts its output into the directory <code>quast_results/result_&lt;DATE&gt;</code> and creates a symlink <code>latest</code> to it inside the directory <code>quast_results/</code>.

<br>
<br>
    QUAST output contains:
<table style="margin-left: 20px; font-size: 1em;">
    <tr>
        <td style="padding-right: 20px;">report.txt</td>
        <td>an assessment summary in a simple text format,</td>
    </tr>
    <tr>
        <td style="padding-right: 20px;">report.tsv</td>
        <td>a tab-separated version of the summary, suitable for spreadsheets (Google Docs, Excel, etc),</td>
    </tr>
    <tr>
        <td style="padding-right: 20px;">report.tex</td>
        <td>a LaTeX version of the summary,</td>
    </tr>
    <tr>
        <td style="padding-right: 20px;">plots.pdf</td>
        <td>plots (file is created if the matplotlib python library is installed),</td>
    </tr>
    <tr>
        <td style="padding-right: 20px;">report.html</td>
        <td>an HTML version of the report with all plots inside it,</td>
    </tr>
    <tr>
        <td style="padding-right: 20px;">contigs_reports/</td>
        <td></td>
    </tr>
    <tr>
        <td style="padding-right: 20px; padding-left: 20px;">misassemblies_report</td>
        <td>a detailed report on misassemblies. See <a href="#sec3.1.2">section 3.1.2</a> for details,</td>
    </tr>
    <tr>
        <td style="padding-right: 20px; padding-left: 20px;">unaligned_report</td>
        <td>a detailed report on unaligned and partially unaligned contigs. See <a href="#sec3.1.3">section 3.1.3</a> for details.</td>
    </tr>
</table>

<br>

Note:
<ul style="margin-top: -12px;">
    <li>metrics based on a reference genome are computed only if a reference is provided (see  <a href="#sec1.3">section 2.4</a>), </li>
    <li>metrics based on genes and operons are computed only if proper annotations are provided (see <a href="#sec1.3">section 2.4</a>). </li>
</ul>
</p>


<div class='metrics_description'>
<a name="sec3.1"></a>
<h3 style='margin-bottom: -15px;'>3.1 Metrics description</h3>
<a name="sec3.1.1"></a>
<h4>3.1.1 Summary report</h4>

<p><span class='metric-name'># contigs (&ge;<span class="rhs">&nbsp;</span>x<span class="rhs">&nbsp;</span>bp)</span> 
    is total number of contigs of length <code>&ge; x<span class="rhs">&nbsp;</span>bp</code>.
    Not affected by the <code>&#8209;&#8209;min&#8209;contig</code> parameter (see <a href="#sec2.4">section 2.4</a>).</p>

<p><span class='metric-name'>Total length (&ge;<span class="rhs">&nbsp;</span>x<span class="rhs">&nbsp;</span>bp)</span> 
    is the total number of bases in contigs of length <code>&ge; x<span class="rhs">&nbsp;</span>bp</code>.
    Not affected by the <code>&#8209;&#8209;min&#8209;contig</code> parameter (see <a href="#sec2.4">section 2.4</a>).<br>
    <br>
    <i>All remaining metrics are computed only the contigs that exceed the threshold specified by specified by the
    <code>&#8209;&#8209;min&#8209;contig</code> option (see <a href="#sec2.4">section 2.4</a>, default is 500).</i>
</p>

<p><span class='metric-name'># contigs</span> is the total number of contigs in the assembly.</p>

<p><a name='largest_contig'></a><span class='metric-name'>Largest contig</span> is the length of the longest contig in the assembly.</p>

<p><span class='metric-name'>Total length</span> is the total number of bases in the assembly.</p>

<p><span class='metric-name'>Reference length</span> is the total number of bases in the reference.</p>

<p><a name='N50'></a><span class='metric-name'>N50</span> is the length for which the collection of all contigs of that length or longer 
    covers at least half an assembly.<br>

<p><a name='NG50'></a><span class='metric-name'>NG50</span> is the length for which the collection of all contigs of that length or longer 
    covers at least half a reference genome.<br> This metric is computed only if
    a reference genome is provided.</p>

<p><span class='metric-name'>N75 and NG75</span> are defined similarly with 75<span class="rhs">&nbsp;</span>% instead of 50<span class="rhs">&nbsp;</span>%.</p>

<p><span class='metric-name'>L50 (L75, LG50, LG75)</span> is the number of contigs as long as N50 (N75, NG50, NG75)<br>
    In other words, L50, for example, is the minimal number of contigs that cover half the assembly.</p>

<p><span class='metric-name'># misassemblies</span> is the number of positions in the contigs that satisfy one of the following criteria:<br>
    <ul style='margin-top: -24px;'>
    <li>the left flanking sequence aligns over 1<span class="rhs">&nbsp;</span>kbp away from the right flanking sequence on the reference;
    <li>flanking sequences overlap on more than 1<span class="rhs">&nbsp;</span>kbp;
    <li>flanking sequences align to different strands or different chromosomes. 
    </ul>
    This metric requires a reference genome.</p>

<p><span class='metric-name'># misassembled contigs</span> is the number of contigs that contain misassembly events.</p>

<p><span class='metric-name'>Misassembled contigs length</span> is the total number of bases in misassembled contigs.</p>

<p><span class='metric-name'># unaligned contigs</span> is the number of contigs that have no alignment to the
    reference sequence. The value "X<span class='rhs'>&nbsp;</span>+<span class='rhs'>&nbsp;</span>Y part" means X totally unaligned contigs plus Y partially unaligned contigs. </p>

<p><span class='metric-name'>Unaligned contigs length</span> is the total length of all unaligned regions in the assembly. </p>

<!--p><span class='metric-name'># ambiguous contigs</span> is the number of contigs which have reference alignments
    of equal quality in multiple locations on the reference. </p>

<p><span class='metric-name'>Ambiguous contigs length</span> is the number of total bases contained in all ambiguous contigs. </p-->

<p><span class='metric-name'>Genome fraction (%)</span> is the percentage of alinged bases in the reference.
    A base in the reference is aligned if there is at least one contig with at least one alignment to this base. 
    Contigs from repetitive regions may map to multiple places, and thus may be counted multiple times.</p>

<p><span class='metric-name'>Duplication ratio</span> is the total number of aligned bases in the assembly divided by the total number of aligned bases in the 
    reference (see <span class='metric-ref'>Genome fraction (%)</span> for the 'aligned base' defenition). If the assembly contains many contigs that cover the same 
    regions of the reference, its <span class='metric-ref'>duplication ratio</span> may be much larger than 1. This may occur due to overestimating 
    repeat multiplicities and due to small overlaps between contigs, among other reasons.</p>

<p><a name='GC'></a><span class='metric-name'>GC (%)</span> is the total number of G and C nucleotides in the assembly,
    divided by the total length of the assembly.</p>

<p><span class='metric-name'>Reference GC (%)</span> is the percentage of G and C nucleotides in the reference.</p>

<p><span class='metric-name'># mismatches per 100<span class="rhs">&nbsp;</span>kbp</span> is the average number of mismatches 
    per 100<span class='hs'></span>000 aligned bases. True SNPs and sequencing errors are not distinguished and are counted equally.
    </ul>
</p>

<p><span class='metric-name'># indels per 100<span class="rhs">&nbsp;</span>kbp</span> is the average number of indels per 100<span class='hs'></span>000 aligned bases. 
    Several consecutive single nucleotide indels are counted as one indel.</p>

<p><span class='metric-name'># N's per 100<span class="rhs">&nbsp;</span>kbp</span> is the average number of uncalled bases (N's) per 100<span class='hs'></span>000 assembly bases.</p>

<p><a name='genes'></a><span class='metric-name'># genes</span> is the number of genes in the assembly (complete and partial), based on a user-provided 
    list of gene positions in the reference. A gene 'partially covered' if the assembly contains at least 100<span class="rhs">&nbsp;</span>bp 
    of this gene but not the whole one.<br>
    <span style='line-height: 50%;'>&nbsp;</span><br>
    This metric is computed only if a reference genome and an annotated list of gene positions are provided (see <a href="#sec2.4">section 2.4</a>).</p>

<p><span class='metric-name'># operons</span> is defined similarly to <span class='metric-ref'># genes</span>, but an operon positions file required instead.</p>

<p><a name='predicted_genes'><span class='metric-name'># predicted genes</span> is the number of genes in the assembly
    found by GeneMarkS, GlimmerHMM or MetaGeneMark. See the description of the <a href='#gene_finding'><code>--gene-finding</code></a> option for details.</p>

<p><span class='metric-name'>Largest alignment</span> is the length of the largest continuous alignment in the assembly.
    A value can be smaller than a value of <a href='largest_contig'><span class='metric-ref'>largest contig</span></a> if the largest contig is misassembled.</p>

<p><a name='NAx'><a name='NGAx'></a><span class='metric-name'>NA50, NGA50, NA75, NGA75, LA50, LA75, LGA50, LGA75</span> ("A" stands for "aligned") are similar to
    the corresponding metrics without "A", but in this case aligned blocks instead of contigs are considered.<br>
    Aligned blocks are obtained by breaking contigs in misassembly events and removing all analigned bases.</p>


<a name="sec3.1.2"></a>
<h4>3.1.2 Misassemblies report</h4>

<p><span class='metric-name'># misassemblies</span> is the same as <span class='metric-ref'># misassemblies</span> from <a href="#sec3.1.1">section 3.1.1</a>.
    However, this report also contains a classification of all misassemblies into three groups: <span class='metric-ref'>relocations</span>, <span class='metric-ref'>translocations</span>,
    and <span class='metric-ref'>inversions</span> (see below).</p>

<p><span class='metric-name'>Relocation</span> is a misassembly where the left flanking sequence aligns over 1<span class="rhs">&nbsp;</span>kbp away from the right flanking
sequence on the reference, or they overlap by more than 1<span class="rhs">&nbsp;</span>kbp, and both flanking sequences align on the same chromosome. </p>

<p><span class='metric-name'>Translocation</span> is a misassembly where the flanking sequences align on different chromosomes. </p>

<p><span class='metric-name'>Inversion</span> is a misassembly where the flanking sequences align on
opposite strands of the same chromosome. </p>

<p><span class='metric-name'># misassembled contigs</span> and <span class='metric-ref'>misassembled contigs length</span> are the same as the metrics from 
    <a href="#sec3.1.1">section 3.1.1</a> and are counted among all contigs with any type of a misassembly 
    (relocation, translocation or inversion). </p>

<p><span class='metric-name'># local misassemblies</span> is the number of breakpoints that satisfy the following conditions: 
<ol class='my_ol' style='margin-top: -24px;'>
<li>Two or more distinct alignments cover the breakpoint.</li>
<li>The gap between left and right flanking sequences is less than 1<span class="rhs">&nbsp;</span>kbp.</li>
<li>The left and right flanking sequences both are on the same strand of the same chromosome of the reference genome.</li>
</ol> </p>

<p><span class='metric-name'># mismatches</span> is the number of mismatches in all aligned bases.</p>

<p><span class='metric-name'># indels</span> is the number of indels in all aligned bases.</p>

<p><span class='metric-name'># short indels (&le;<span class="rhs">&nbsp;</span>5<span class="rhs">&nbsp;</span>bp)</span> is the number of indels of length <code>&le;<span class="rhs">&nbsp;</span>5<span class="rhs">&nbsp;</span>bp</code>.</p>

<p><span class='metric-name'># long indels (&gt;<span class="rhs">&nbsp;</span>5<span class="rhs">&nbsp;</span>bp)</span> is the number of indels of length <code>&gt;<span class="rhs">&nbsp;</span>5<span class="rhs">&nbsp;</span>bp</code>.</p>

<p><span class='metric-name'>Indels length</span> is the total number of bases contained in all indels.</p>


<a name="sec3.1.3"></a>
<h4>3.1.3 Unaligned report</h4>

<p><span class='metric-name'># fully unaligned contigs</span> is the number of contigs that have no alignment to the reference sequence.</p>

<p><span class='metric-name'>Fully unaligned length</span> is the total number of bases in all unaligned contigs.</p>

<p><span class='metric-name'># partially unaligned contigs</span> is the number of contigs that are not fully unaligned, but have fragments
    with no alignment to the reference sequence.</p>

<p><span class='metric-name'># with misassembly</span> is the number of partially unaligned contigs that have a misassembly in their aligned fragment.
    Note that such misassemblies are not counted in <span class='metric_ref'># misassemblies</span> and other <span class='metric_ref'>misassemblies</span> statistics.</p>

<p><span class='metric-name'># both parts are significant</span> is the number of partially unaligned contigs that have both aligned and unaligned fragments
    longer than the value of <a href='#min_contig'><code>--min-contig</code></a>.

<p><span class='metric-name'>Partially unaligned length</span> is the total number of unaligned bases in all partially unaligned contigs.</p>

<p><span class='metric-name'># N's</span> is the total number of uncalled bases (N's) in the assembly.</p>


<a name="sec3.2"></a>
<h3>3.2 Plots description</h3>

<p><span class='metric-name'>Cumulative length plot</span> shows the growth of contig lengths. On the x-axis, contigs are ordered from the largest 
    to smallest. The y-axis gives the size of the x largest contigs in the assembly.</p>

<p><span class='metric-name'>Nx plot</span> shows <a href='#N50'><span class='metric-ref'>Nx</span></a> values as x varies from 0 to 100<span class='rhs'>&nbsp;</span>%.</p>

<p><span class='metric-name'>NGx plot</span> shows <a href='#NG50'><span class='metric-ref'>NGx</span></a> values as x varies from 0 to 100<span class='rhs'>&nbsp;</span>%.</p>

<p><span class='metric-name'>GC content plot</span> shows the distribution of <a href='#GC'>GC content</a> in the contigs.<br>
    <span style='line-height: 50%;'>&nbsp;</span><br>
    The x value is the GC percentage (0 to 100<span class='rhs'>&nbsp;</span>%).<br>
    The y value is the number of non-overlapping 100<span class="rhs">&nbsp;</span>bp windows which GC content equals x<span class='rhs'>&nbsp;</span>%.<br>
    <span style='line-height: 50%;'>&nbsp;</span><br>
    For a single genome, the distribution is typically Gaussian. However, for assemblies with contaminants, the GC distribution 
    appears to be a superposition of Gaussian distributions, giving a plot with multiple peaks.
</p>

<p><span class='metric-name'>Cumulative length plot for aligned contigs</span> shows the growth of lengths of aligned blocks.
    If a contig has a misassembly, QUAST breaks it into smaller pieces called aligned blocks.<br>
    <span style='line-height: 50%;'>&nbsp;</span><br>
    On the x-axis, blocks are ordered from the largest 
    to smallest. The y-axis gives the size of the x largest aligned blocks.<br>
    This plot is created only if a reference genome is provided.<p>

<p><span class='metric-name'>NAx and NGAx plots</span><br>
    These plots are similar to the <span class='metric-ref'>Nx</span> and <span class='metric-ref'>NGx</span> plots but for the <a href='#NAx'>NAx and NGAx</a> metrics respectively.
    These plots are created only if a reference genome is provided.</p>

<p><a name='gene_plot'></a><span class='metric-name'>Genes plot</span> shows the growth rate of full genes in assemblies.<br>
    The y-axis is the number of full genes in the assembly, and the x-axis is the number of contigs in the assembly (from the largest one to the smallest one).<br>
    This plot could be created only if a <a href='#sec2.2'>reference and genes annotations files</a> are given.</p>

<p><span class='metric-name'>Operons plot</span> is similar to the previous one but for operons.</p>

</div>

<a name="sec4"></a><p>
<h2>4. Adjusting QUAST reports and plots</h2>
<p> You can easily change content, order of metrics, and metric names in all QUAST reports. For doing this,
    please edit the <code>CONFIGURABLE PARAMETERS</code> section in <code>libs/reporting.py</code>. It contains a lot of informative comments,
    which will help you to adjust QUAST reports easily even if you are new to Python.
</p>    

<p> You can also adjust plot colors, style and width of lines, legeng font, plots output format, etc.
    Please see the <code>CONFIGURABLE PARAMETERS</code> section in <code>libs/plotter.py</code>.
</p>

<p> Note: if you restart QUAST on the same directory with new parameters, is will reuse alignments and run much faster.
    See the description of the <code>-o</code> option in <a href="#sec2.4">section 2.4</a>.
</p>

<a name="sec5"></a><p>
<h2>5. Citation</h2><br>
If you use QUAST in your research, please include <i>Gurevich et al., 2013</i> into your reference list:<br>
<div style='margin-top: 5px; padding-left: 30px; padding-top: 5px; padding-bottom: 5px; background-color: #EEE'>
Alexey Gurevich, Vladislav Saveliev, Nikolay Vyahhi and Glenn Tesler, <br>
QUAST: quality assessment tool for genome assemblies, <br>
Bioinformatics (2013) 29 (8): 1072-1075. doi: <a href="http://dx.doi.org/10.1093/bioinformatics/btt086">10.1093/bioinformatics/btt086</a><br>
First published online: February 19, 2013
</div>

<a name="sec6"></a><p>
<h2>6. Feedback and bug reports</h2>
<p> We will be thankful if you help us make QUAST better by sending your comments, bug reports, and suggestions
    to <a href="mailto:quast.support@bioinf.spbau.ru">quast.support@bioinf.spbau.ru</a>.
</p>

<p>
    We kindly ask you to attach the <code>quast.log</code> file from output directory (or an entire archive of the folder) if you have troubles running QUAST.
    <br><br>
    Note that if you didn't specify the output directory manually, it is going to be automatically set to <code>quast_results/results_&lt;date_time&gt;</code>, with a symbolic link <code>quast_results/latest</code> to that directory.
</p>

<br>
<br>
</body>
</html>
