<!DOCTYPE html>
<html lang="en">
<head>
    <script type="text/javascript" src="report_html_files/jquery-1.7.2.min.js"></script>
    <script type="text/javascript" src="report_html_files/flot/jquery.flot.js"></script>
    <script type="text/javascript" src="report_html_files/flot/jquery.flot.dashes.js"></script>
    <!–[if lte IE 8]><script type="text/javascript" src="report_html_files/flot/excanvas.min.js"></script><![endif]–>

    <script type="text/javascript" src="report_html_files/scripts/utils.js"></script>
    <script type="text/javascript" src="report_html_files/scripts/build_total_report.js"></script>
    <script type="text/javascript" src="report_html_files/scripts/draw_cumulative_plot.js"></script>
    <script type="text/javascript" src="report_html_files/scripts/draw_nx_plot.js"></script>
    <script type="text/javascript" src="report_html_files/scripts/draw_gc_plot.js"></script>
    <script type="text/javascript" src="report_html_files/scripts/draw_genes_plot.js"></script>
    <script type="text/javascript" src="report_html_files/scripts/build_report.js"></script>

    <link rel="stylesheet" href="report_html_files/bootstrap/bootstrap.min.css"/>
    <!-- Le HTML5 shim, for IE6-8 support of HTML5 elements -->
    <!--[if lt IE 9]>
        <script type="text/javascript" src="report_html_files/ie_html5.js"></script>
    <![endif]-->

    <script type="text/javascript">
        $(document).ready(buildReport);
    </script>

    <style type="text/css">
        body {
            padding-bottom: 30px;
        }
        .content {
            margin-left: 120px;
            padding-top: 50px;
            clear: both;
        }
        #header {
            display: block;
            width: 100%;
            margin-bottom: 2em;
            font-size: 1.5em;
        }
        #subheader { /* not used in Quast html report */
            margin-top: -13px;
            margin-bottom: 20px;
        }
    </style>

    <link rel="stylesheet" type="text/css" href="report_html_files/common.css" />
    <link rel="stylesheet" type="text/css" href="report_html_files/report.css" />
</head>

<body>
    <div class="content">
        <div id='header'>
            <span class='smallcapitals'>QUAST</span> report<br><br>
            <span style='font-size: 0.7em;'>
            <span id='subheader'>Loading...</span>
            <span id='mincontig'></span>
        	</span>
        </div>

        <noscript>
            <div style="margin-left: 0px;">
                Unfortunately, <span style="font-weight: bold">JavaScript</span> in your
                browser <span style="font-weight: bold">is disabled</span> or is not supported.</br>
                We need JavaScript to build report and plots.
            </div>
        </noscript>

        <div class='blocks'>
            <div id='report'>
            </div>
            <div class='plot-wrapper' id='cumulative-plot-div'>
            </div>
            <div class='plot-wrapper' id='nx-plot-div'>
            </div>
            <div class='plot-wrapper' id='ngx-plot-div'>
            </div>
            <div class='plot-wrapper' id='nax-plot-div'>
            </div>
            <div class='plot-wrapper' id='ngax-plot-div'>
            </div>
            <div class='plot-wrapper' id='genes-plot-div'>
            </div>
            <div class='plot-wrapper' id='operons-plot-div'>
            </div>
            <div class='plot-wrapper' id='gc-plot-div'>
            </div>
        </div>
    </div>


    <div class='json-code'>
        <div id='report-json'>
            {"date": "07 October 2012, Sunday, 01:33:51", "header": ["Assembly", "# contigs (>= 500 bp)", "# contigs (>= 1000 bp)", "Total length (>= 500 bp)", "Total length (>= 1000 bp)", "# contigs", "Largest contig", "Total length", "Reference length", "GC (%)", "N50", "L50", "NG50", "LG50", "N75", "L75", "NG75", "LG75", "# misassemblies", "# misassembled contigs", "Misassembled contigs length", "# unaligned contigs", "Unaligned contigs length", "# ambiguous contigs", "Ambiguous contigs length", "Genome fraction (%)", "Duplication ratio", "Reference GC (%)", "N's (%)", "# mismatches per 100 kbp", "# indels per 100 kbp", "# genes", "NA50", "LA50", "NGA50", "LGA50", "NA75", "LA75", "NGA75", "LGA75"], "results": [["SPAdes_contigs_1", 1, 1, 1000, 1000, 1, 1000, 1000, 1000, 50.700000000000003, 1000, 1, 1000, 1, 1000, 1, 1000, 1, 0, 0, 0, "0 + 0 part", 0, 0, 0, 100.0, 1.0, 50.700000000000003, 0.0, 0.0, 0.0, "1 + 0 part", 1000, 1, 1000, 1, 1000, 1, 1000, 1], ["SPAdes_contigs_2", 1, 0, 760, 0, 1, 760, 760, 1000, 51.049999999999997, 760, 1, 760, 1, 760, 1, 760, 1, 0, 0, 0, "0 + 0 part", 0, 0, 0, 76.0, 1.0, 50.700000000000003, 0.0, 0.0, 0.0, "1 + 0 part", 1360, 1, 1360, 1, 1360, 1, 1360, 1]], "min_contig": 200}
        </div>
        <div id="contigs-lengths-json">
            {"lists_of_lengths": [[1000], [760]], "filenames": ["SPAdes_contigs_1", "SPAdes_contigs_2"]}
        </div>
        <div id='aligned-contigs-lengths-json'>
            {"lists_of_lengths": [[1000], [1360]], "filenames": ["SPAdes_contigs_1", "SPAdes_contigs_2"]}
        </div>
        <div id='assemblies-lengths-json'>
            {"assemblies_lengths": [1000, 760], "filenames": ["SPAdes_contigs_1", "SPAdes_contigs_2"]}
        </div>
        <div id='reference-length-json'>
            {"reflen": 1000}
        </div>
        <div id='contigs-json'>
            {"contigs": {"SPAdes_contigs_1": [[[1, 1000]]], "SPAdes_contigs_2": [[[1, 600], [841, 1000]]]}, "filenames": ["SPAdes_contigs_1", "SPAdes_contigs_2"]}
        </div>
        <div id='genes-json'>
            {"found": [1], "genes": [[190, 220]]}
        </div>
        <div id='operons-json'>
            {{ operons }}
        </div>
        <div id='gc-json'>
            {"lists_of_gc_info": [[[1, 42.0], [1, 32.0], [1, 56.0], [1, 47.0], [1, 55.0], [1, 61.0], [1, 47.0], [1, 51.0], [1, 58.0], [1, 58.0]], [[1, 42.0], [1, 32.0], [1, 56.0], [1, 47.0], [1, 55.0], [1, 61.0], [1, 59.0]]], "filenames": ["SPAdes_contigs_1", "SPAdes_contigs_2"]}
        </div>
    </div>

    <div class='json-code'>
        <div id='glossary-json'>
            {
    "N50"                           : "<span>is the contig length such that using longer or equal length contigs produces half (50%) the bases of the assembly. Usually there is no value that produces exactly 50%, so the more technical definition is the minimal length x such that using contigs of length at least x accounts for at least 50% of the total assembly length.</span>",
    "NG50"                          : "<span>is the contig length such that using longer or equal length contigs produces half (50%) the bases of the reference genome. This metric could be computed only if the reference is given.</span>",
    "N75"                           : "<span>is the contig length such that using longer or equal length contigs produces 75% the bases of the assembly. Usually there is no value that produces exactly 75%, so the more technical definition is the minimal length x such that using contigs of length at least x accounts for at least 75% of the total assembly length.</span>",
    "NG75"                          : "<span>is the contig length such that using longer or equal length contigs produces 75% the bases of the reference genome. This metric could be computed only if the reference is given.</span>",

    "L50"                           : "<span>is the minimal number of contigs that summarily cover half (50%) the bases of the assembly. In other words, it's the number of contigs of lenght at least N50.</span>",
    "LG50"                          : "<span>is the minimal number of contigs that summarily cover half (50%) the bases of the reference genome. In other words, it's the number of contigs of lenght at least NG50. This metric could be computed only if the reference is given.</span>",
    "L75"                           : "<span>is the minimal number of contigs that summarily cover 75% the bases of the assembly. In other words, it's the number of contigs of lenght at least N75.</span>",
    "LG75"                          : "<span>is the minimal number of contigs that summarily cover 75% the bases of the reference genome. In other words, it's the number of contigs of lenght at least NG75. This metric could be computed only if the reference is given.</span>",

    "NA50"                          : "<span>is N50 where the lengths of aligned blocks are counted instead of contigs lengths. I.e. if a contig has a misassembly with respect to the reference, the contig is broken into smaller pieces.</span>",
    "NGA50"                         : "<span>is NG50 where the lengths of aligned blocks are counted instead of contigs lengths. I.e. if a contig has a misassembly with respect to the reference, the contig is broken into smaller pieces.</span>",
    "NA75"                          : "<span>is N75 where the lengths of aligned blocks are counted instead of contigs lengths. I.e. if a contig has a misassembly with respect to the reference, the contig is broken into smaller pieces.</span>",
    "NGA75"                         : "<span>is NG75 where the lengths of aligned blocks are counted instead of contigs lengths. I.e. if a contig has a misassembly with respect to the reference, the contig is broken into smaller pieces.</span>",

    "LA50"                          : "<span>is L50 where aligned blocks are counted instead of contigs. I.e. if a contig has a misassembly with respect to the reference, the contig is broken into smaller pieces.</span>",
    "LGA50"                         : "<span>is LG50 where aligned blocks are counted instead of contigs. I.e. if a contig has a misassembly with respect to the reference, the contig is broken into smaller pieces.</span>",
    "LA75"                          : "<span>is L75 where aligned blocks are counted instead of contigs. I.e. if a contig has a misassembly with respect to the reference, the contig is broken into smaller pieces.</span>",
    "LGA75"                         : "<span>is LG75 where aligned blocks are counted instead of contigs. I.e. if a contig has a misassembly with respect to the reference, the contig is broken into smaller pieces.</span>",

    "Average %IDY"                  : "<span>is an average of alignment identity percent (Nucmer measure of alignment accuracy) among all contigs.</span>",
    "# misassemblies"               : "<span>is the number of positions at the assembled contigs where the left flanking sequence aligns over 1kb away from the right flanking sequence on the reference or they overlap on more than 1 kb or flanking sequences align on different strands.</span>",
    "# misassembled contigs"        : "<span>is the number of contigs that contain misassembly events.</span>",
    "Misassembled contigs length"   : "<span>is the number of total bases contained in all contigs that have one or more misassemblies.</span>",
    "Misassembled and unaligned"    : "<span>is the number of contigs which contain misassembly events but which have alignment to the reference of length less than 10% of contigs length (almost unaligned contigs).</span>",
    "# unaligned contigs"           : "<span>is the number of contigs that have no alignment to the reference sequence. The value 'X + Y part' means: X totally unaligned contigs plus Y partially unaligned contigs.</span>",
    "Unaligned contigs length"      : "<span>is the total number of unaligned bases in totally and partially unaligned contigs (only bases from unaligned parts).</span>",
    "# ambiguous contigs"           : "<span>is the number of contigs which have reference alignments of equal quality in multiple locations on the reference.</span>",
    "Ambiguous contigs length"      : "<span>is the number of total bases contained in all ambiguous contigs.</span>",
    "Genome fraction (%)"           : "<span>is the total number of aligned bases in the reference, divided by the genome size. A base in the reference genome is counted as aligned if there is at least one contig with at least one alignment to this base. Contigs from repeat regions may map to multiple places, and thus may be counted multiple times in this quantity.</span>",
    "GC (%)"                        : "<span>is the total number of G and C nucleotides in the assembly, divided by the total length of the assembly.</span>",
    "Reference GC (%)"              : "<span>is the total number of G and C nucleotides in the reference, divided by the total length of the reference.</span>",
    "# mismatches per 100 kbp"      : "<span>is an average number of mismatches per 100000 aligned bases.</span>",
    "# genes"                       : "<span>is the number of genes in the assembly (complete and partial), based on a user-provided annotated list of gene positions in the reference genome. A gene counts as 'partially covered' if the assembly contains at least 100 bp of this gene but not the whole gene.</span>",
    "# operons"                     : "<span>is the number of operons in the assembly (complete and partial), based on a user-provided annotated list of operon positions in the reference genome. An operon counts as 'partially covered' if the assembly contains at least 100 bp of this operon but not the whole operon.</span>",

    "Cumulative length"             : "<span>plot shows the growth rate of assemblies lengthes. The y-axis is the size of assembly and x-axis is the number of contigs in the assembly (from the largest one to the smallest one).</span>",
    "Nx"                            : "<span>plot shows the changes of Nx metric value with the percentage of an assembly length</span>  <span>Nx is the minimal length <b>y</b> such that using contigs of length at least <b>y</b> accounts for at least x% of the total assembly length.</span>",
    "NGx"                           : "<span>plot shows the changes of NGx metric value with the percentage of a reference length</span> <span>NGx is the contig length such that using longer or equal length contigs produces x% the bases of the reference genome. This metric could be computed only if the reference is given.</span>",
    "NAx"                           : "<span>plot shows the changes of NAx metric value with the percentage of an assembly length</span> <span>NAx is Nx with usage the lengths of aligned blocks instead of contigs lengths.</span>",
    "NGAx"                          : "<span>plot shows the changes of NGAx metric value with the percentage of a reference length</span><span>NGAx is NGx with usage the lengths of aligned blocks instead of contigs lengths.</span>",
    "GC content"                    : "<span>plot shows the distribution of GC percentage among the contigs, i.e., the total number of bp in contigs with such GC content. Typically, the distribution is similar to the Gaussian one. However, in case of contaminants with different GC content the distribution can have more than one curve.</span>"
}
        </div>
    </div>

    <script type="text/javascript">
        $(document).ready(function () {
            $("[rel=tooltip]").tooltip({
                animation: false,
            });
        });
    </script>

    <script type="text/javascript" src="report_html_files/bootstrap/bootstrap-tooltip-5px-lower.min.js"></script>
</body>
</html>
