<!DOCTYPE html>
<html lang="en">
<head>
    <title>QUAST</title>
    <meta http-equiv="Content-Type" content="text/html;charset=utf-8" >

    <link rel="stylesheet" type="text/css" href="quast.css">

    <script src="jquery-1.8.2.js"></script> 
    <script> 
    $(function(){
      $("#common_header").load("common_header.html"); 
    });
    </script>  
</head>

<body>
    <div id="common_header"></div>

    <div class="content">
        <div id='header'>
            <h2>MetaQUAST</h2>
        </div>

            <div>MetaQUAST evaluates and compares metagenome assemblies based on alignments to close references. It is based on <a href="quast.html">QUAST</a> genome quality assessment tool, but addresses features specific for metagenome datasets:</div><ul style="padding: 10px"><li><em>Huge species diversity</em> – the tool accepts multiple references and makes multi-genome tables and plots, including Krona charts.</li><li><em>Commonly unknown species content</em> – the tool automatically detects and downloads reference sequences from NCBI.</li><li><em>Presence of highly relative genomes</em> – the tool detects chimeric contigs and reports "interspecies misassemblies" in addition to the regular assembly errors types.&nbsp;</li></ul><div>&nbsp;</div><div>MetaQUAST can be fed with multiple assemblies, thus is perfect for comparison.</div><div>&nbsp;</div><div>MetaQUAST is distributed <strong>within</strong> QUAST package since version 2.2.</div>
            <p>
                <a href="https://sourceforge.net/projects/quast/files/latest/download" rel="nofollow"><img alt="Download QUAST" src="https://img.shields.io/sourceforge/dm/quast.svg"></a>
            </p>

            <p>MetaQUAST specific sections in <a href="http://quast.bioinf.spbau.ru/manual.html">the Manual</a>: <b><a href="http://quast.bioinf.spbau.ru/manual.html#sec2.5">running</a>, <a href="http://quast.bioinf.spbau.ru/manual.html#sec3.3">output description</a></b></p>

        <p><strong>Citation:</strong>
        <div style='margin-top: 5px; padding-right: 30px; padding-left: 30px; padding-top: 5px; padding-bottom: 5px; background-color: #EEE; display: inline-block'>
            Alla Mikheenko, Vladislav Saveliev, Alexey Gurevich, <br>
            MetaQUAST: evaluation of metagenome assemblies, <br>
            <i>Bioinformatics</i> (2016) 32 (7): 1088-1090. doi: <a href="http://dx.doi.org/10.1093/bioinformatics/btv697">10.1093/bioinformatics/btv697</a><br>
            First published online: November 26, 2015
        </div>
        </p>

        <a href="http://bioinformatics.oxfordjournals.org/citmgr?gca=bioinfo%3Bbtv697v1" style="color: rgb(75, 92, 195);">Citation of other formats</a>

        <div style="height: 5px;">&nbsp;</div>
        <p><strong>Examples of comparison of 4 assemblers on 3 datasets generated with MetaQUAST v3.2:</strong></p><div><a href="http://quast.bioinf.spbau.ru/static/metaquast/cami/report.html">CAMI simulated toy dataset</a>&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;<a href="http://quast.bioinf.spbau.ru/static/metaquast/metahit/report.html">MetaHIT real dataset&nbsp;</a>&nbsp; &nbsp;&nbsp;<a href="http://quast.bioinf.spbau.ru/static/metaquast/hmp/report.html">HMP real dataset</a></div><div>&nbsp;</div><div><span style="font-size:16px;"><a href="img/metahit_screenshot_v3.2.png"><img alt="" src="http://bioinf.spbau.ru/sites/default/files/metahit_screenshot_v3.2.png" style="width: 1020px; height: 638px; border-width: 1px; border-style: solid;"></a></span></div>

        <br>
        <div><em>Please help us to make MetaQUAST better by sending your comments, bug reports, and suggestions to&nbsp;<a href="mailto:quast.support@bioinf.spbau.ru">quast.support@bioinf.spbau.ru</a></em></div>
    </div>
</body>
</html>
