{% extends "base-quast.html" %}

{% block title %}About QUAST{% endblock %}

{% block extra_head %}
    <link rel="stylesheet" href="/static/docs.css" />
{% endblock %}

{% block content %}
    <h2><a name="top"></a>About</h2>
    <h3><a href="#intro">Short intro</h3><br />
    <a href="#further">Further reading</a><br />
    <a href="#citations">How to cite</a><br />
    <a href="#history">Brief project history</a><br />
    <a href="#authors">Authors</a><br />
    <a href="#funding">Funding</a><br />
    <a href="#license">License and Terms of use</a></h3>
    <h3><a name="intro"></a> Short intro</h3>
    <p><strong>QUAST evaluates genome assemblies by computing various metrics, including</strong></p>
    <ul>
    <li>N50, length for which the collection of all contigs of that length or longer covers at least 50% of assembly length</li>
    <li>NG50, where length of the reference genome is being covered</li>
    <li>NA50 and NGA50, where aligned blocks instead of contigs are taken</li>
    <li>misassemblies, misassembled and unaligned contigs or contigs bases</li>
    <li>genes and operons covered</li>
    </ul>
    <p><strong>.. and builds convenient plots for different metrics</strong></p>
    <ul>
    <li>cumulative contigs length</li>
    <li>all kinds of N-metrics</li>
    <li>genes and operons covered</li>
    <li>GC content</li>
    </ul>
    <p><strong>QUAST can work with and without a reference genome</strong>, though the output is considerably smaller in the latter case. The pipeline relies on multiple open-source tools, including <a href="https://github.com/lh3/minimap2">minimap2</a>, <a href="https://ccb.jhu.edu/software/glimmerhmm/">GlimmerHMM</a>, and <a href="https://busco.ezlab.org/">BUSCO</a>.</p>
    <h3><a name="further"></a> Further reading</h3>
    <p>More details are on the <a href="http://cab.spbu.ru/software/quast">project page</a> and in <a href="http://dx.doi.org/10.1093/bioinformatics/btt086">Gurevich et al., Bioinformatics, 2013</a>.<br />
    You may also want to check out the <a href="help.md">Help</a> page with a step-by-step tutorial on using this website and interpreting its output.<br />
    Thorough metric descriptions are available in the <a href="http://cab.cc.spbu.ru/quast/manual.html#sec3.1">online manual</a>. A nice illustrated explanation of N50 and similar metrics by Elin Videvall (with assistance from the QUAST team) is <a href="https://www.molecularecologist.com/2017/03/29/whats-n50/">here</a>.<br />
    Finally, we invite you to read all QUAST publications listed <a href="#citations">below</a> to understand the algorithms behind the tool and its applicability to various domains.</p>
    <p>Don’t hesitate to <a href="contact.md">contact us</a> if you still have any questions!</p>

    <h3><a name="citations"></a> How to cite</h3> If you have found QUAST useful in your research, please cite:
    <p><strong>Versatile genome assembly evaluation with QUAST-LG</strong><br />
    Alla Mikheenko, Andrey Prjibelski, Vladislav Saveliev, Dmitry Antipov, Alexey Gurevich<br />
    <em>Bioinformatics</em> (2018) DOI: <a href="https://doi.org/10.1093/bioinformatics/bty266">10.1093/bioinformatics/bty266</a></p>
    <p><strong>Icarus: visualizer for de novo assembly evaluation</strong><br />
    Alla Mikheenko, Gleb Valin, Andrey Prjibelski, Vladislav Saveliev, Alexey Gurevich<br />
    <em>Bioinformatics</em> (2016) DOI: <a href="https://doi.org/10.1093/bioinformatics/btw379">10.1093/bioinformatics/btw379</a></p>
    <p><strong>MetaQUAST: evaluation of metagenome assemblies</strong><br />
    Alla Mikheenko, Vladislav Saveliev, Alexey Gurevich<br />
    <em>Bioinformatics</em> (2016) DOI: <a href="https://doi.org/10.1093/bioinformatics/btv697">10.1093/bioinformatics/btv697</a></p>
    <p><strong>QUAST: quality assessment tool for genome assemblies</strong><br />
    Alexey Gurevich, Vladislav Saveliev, Nikolay Vyahhi and Glenn Tesler<br />
    <em>Bioinformatics</em> (2013) DOI: <a href="https://doi.org/10.1093/bioinformatics/btt086">10.1093/bioinformatics/btt086</a></p>
    <h3><a name="history"></a> Brief project history</h3>
    <p><strong>2011, July</strong> — project start<br />
    <strong>2012, August 14</strong> — first public release at <a href="https://sourceforge.net/projects/quast/files/">SourceForge</a> (QUAST v.1.0.0)<br />
    <strong>2012, September 27</strong> — the web server first launch (at http://quast.bioinf.spbau.ru/, shut down in January 2020)<br />
    <strong>2013, February 19</strong> — the first <a href="http://dx.doi.org/10.1093/bioinformatics/btt086">QUAST publication</a> was out in <em>Bioinformatics</em><br />
    <strong>2013, April 15</strong> — Prof. Steven Salzberg <a href="https://facultyopinions.com/prime/717981369?bd=1&amp;ui=24116">recommended QUAST</a> at Faculty Opinions (former F1000)<br />
    <strong>2015, July 13</strong> — the <a href="https://github.com/ablab/quast">QUAST repository</a> was opened for public access on GitHub, the tool became fully open sourced<br />
    <strong>2018, November 2</strong> — the total number of QUAST downloads exceeded 100,000 (<a href="https://anaconda.org/bioconda/quast">bioconda</a> &amp; <a href="https://sourceforge.net/projects/quast/files/">SourceForge</a>)<br />
    <strong>2020, April 4</strong> — the web server relauched at its <a href="http://cab.cc.spbu.ru/quast/">current location</a><br />
    <strong>2021, August 31</strong> — <a href="https://chanzuckerberg.com/newsroom/czi-awards-16-million-for-foundational-open-source-software-tools-essential-to-biomedicine/">CZI acknowledged</a> QUAST as an essential software tool for biomedicine and supported with a grant (jointly with <a href="https://cab.spbu.ru/software/spades">SPAdes</a>)</p>
    <h3><a name="authors"></a> Authors</h3>
    <p><strong>Core developers</strong> (in alphabetical order):<br />
    Alexey Gurevich (since 2011)<br />
    Alla Mikheenko (since 2015)<br />
    Vlad Saveliev (2012-2017)</p>
    <p><strong>Also contributed</strong> (in alphabetical order):<br />
    Dmitry Antipov<br />
    Aleksey Komissarov<br />
    Andrey Prjibelski<br />
    Glenn Tesler<br />
    Gleb Valin<br />
    Irina Vasilinetc<br />
    Nikolay Vyahhi<br />
    .. and dozens of QUAST users who provided their feedback</p>
    <p><strong>Logo design</strong> by Elena Strelnikova</p>
    <h3><a name="funding"></a> Funding</h3>
    <p>Over the years, the QUAST development was supported by many funding agencies/sources:</p>
    <ul>
    <li>The <a href="https://p220.ru/en/">mega-grant program</a> of the Government of the Russian Federation</li>
    <li><a href="https://rscf.ru/en/">Russian Science Foundation</a></li>
    <li><a href="https://english.spbu.ru/science-4">St. Petersburg State University</a></li>
    <li><a href="https://chanzuckerberg.com/eoss/">The EOSS program</a> of The Chan Zuckerberg Initiative</li>
    </ul>
    <p>The server is running using computational resources provided by the <a href="http://www.cc.spbu.ru/en">Computer Center of Research park</a> of St. Petersburg State University.</p>
    <h3><a name="license"></a> License and Terms of use</h3>
    <p>QUAST is licensed under GPL v2 and it is free to use for everybody; there is no login requirement for the web server. However, some of build-in third-party tools are not under GPL v2. See <a href="http://cab.cc.spbu.ru/quast/LICENSE.txt">LICENSE</a> for details. In particular, the <a href="http://opal.biology.gatech.edu/GeneMark/">GeneMark</a> gene prediction software has a stricter license, so it is not included in the web server pipeline, but can be used with the command-line version if you are eligible.</p>
    <p><strong>Liability</strong>: the QUAST webserver is developed and maintained in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.</p>
    <h3><a href="#top">To the top</a></h3>
{% endblock %}